File: backend/app/services/ai_service.py
import json
import re

from openai import OpenAI

from app.config import settings
from app.schemas import Issue


client = OpenAI(api_key=settings.openai_api_key) if settings.openai_api_key else None


LINTING_PATTERNS: dict[str, list[tuple[str, str, str, str]]] = {
    "python": [
        (r"^\s*print\(", "info", "style", "Use logging instead of print() for production code"),
        (r"except\s*:", "warning", "best-practice", "Bare except clause — catch specific exceptions"),
        (r"^\s*pass\s*$", "info", "style", "Unimplemented block — remove or implement pass"),
        (r"from\s+\w+\s+import\s+\*", "error", "linting", "Wildcard imports pollute namespace — import explicitly"),
        (r"\t+", "warning", "style", "Tabs used for indentation — use spaces (PEP 8)"),
        (r"if\s+.*:\s*$", "info", "linting", "Missing space after comma" if False else None),
    ],
    "javascript": [
        (r"console\.log\(", "info", "style", "Remove console.log in production code"),
        (r"var\s+", "warning", "style", "Use let/const instead of var"),
        (r"==\s", "warning", "best-practice", "Use === instead of =="),
        (r"eval\s*\(", "error", "security", "eval() is dangerous and should never be used"),
        (r"document\.write\(", "warning", "security", "Avoid document.write() — use DOM methods"),
    ],
    "go": [
        (r"error\.New\(fmt\.Sprintf", "info", "style", "Use fmt.Errorf() instead of error.New(fmt.Sprintf())"),
        (r"panic\(", "warning", "best-practice", "Avoid panic outside of init/main — return errors"),
        (r"defer\s+os\.Remove", "info", "linting", "Consider checking Remove error even when deferred"),
    ],
}


def is_ai_available() -> bool:
    return client is not None


def run_local_lint(code: str, language: str) -> list[Issue]:
    issues: list[Issue] = []
    patterns = LINTING_PATTERNS.get(language, [])
    lines = code.split("\n")

    for line_idx, line in enumerate(lines, start=1):
        for pattern, severity, category, message in patterns:
            if re.search(pattern, line):
                issues.append(Issue(
                    line=line_idx, severity=severity, category=category, message=message
                ))

    return issues


def build_review_prompt(code: str, language: str, lint_issues: list[Issue]) -> str:
    lint_section = "\n".join(
        f"- Line {i.line} [{i.severity}] [{i.category}]: {i.message}"
        for i in lint_issues
    ) if lint_issues else "No basic lint issues detected."

    return f"""You are a senior code reviewer. Analyze the following {language} code and provide:

1. **Summary** — 2-3 sentence overview of the code quality
2. **Overall Score** — integer from 0 to 100
3. **Additional Issues** — any issues NOT already listed below (security, performance, bugs, design patterns, edge cases)

Existing lint issues found:
{lint_section}

Code to review:
```{language}
{code}
```

Respond with valid JSON ONLY (no markdown fences):
{{
  "summary": "...",
  "overall_score": 0-100,
  "additional_issues": [
    {{"line": <int or null>, "severity": "error|warning|info", "category": "security|performance|bug|design|best-practice", "message": "...", "suggestion": "..."}}
  ]
}}"""


def call_llm(prompt: str) -> dict | None:
    if not client:
        return None
    try:
        resp = client.chat.completions.create(
            model=settings.openai_model,
            messages=[{"role": "system", "content": "You are a JSON-only code reviewer."}, {"role": "user", "content": prompt}],
            temperature=0.2,
            max_tokens=2000,
        )
        text = resp.choices[0].message.content or ""
        text = text.strip().removeprefix("```json").removeprefix("```").removesuffix("```").strip()
        return json.loads(text)
    except Exception:
        return None


def compute_score(lint_issues: list[Issue], ai_issues: list[Issue]) -> int:
    base = 100
    for i in lint_issues + ai_issues:
        if i.severity == "error":
            base -= 15
        elif i.severity == "warning":
            base -= 8
        else:
            base -= 3
    return max(base, 0)


def review_code(code: str, language: str) -> tuple[list[Issue], str, int]:
    lint_issues = run_local_lint(code, language)
    all_issues = list(lint_issues)
    summary = ""

    ai_result = None
    if is_ai_available():
        prompt = build_review_prompt(code, language, lint_issues)
        ai_result = call_llm(prompt)

    if ai_result:
        summary = ai_result.get("summary", "")
        additional = ai_result.get("additional_issues", [])
        for iss in additional:
            all_issues.append(Issue(
                line=iss.get("line") or 0,
                severity=iss.get("severity", "info"),
                category=iss.get("category", "best-practice"),
                message=iss.get("message", ""),
                suggestion=iss.get("suggestion"),
            ))
        score = ai_result.get("overall_score", compute_score(lint_issues, all_issues[len(lint_issues):]))
    else:
        summary = f"Reviewed {len(lint_issues)} issue(s) via static analysis. Set OPENAI_API_KEY for AI-powered review."
        score = compute_score(lint_issues, [])

    return all_issues, summary, score


LANGUAGE_OPTIONS = [
    {"id": "python", "label": "Python"},
    {"id": "javascript", "label": "JavaScript"},
    {"id": "typescript", "label": "TypeScript"},
    {"id": "go", "label": "Go"},
    {"id": "rust", "label": "Rust"},
    {"id": "java", "label": "Java"},
    {"id": "cpp", "label": "C++"},
    {"id": "ruby", "label": "Ruby"},
]

File: backend/src/services/aiService.ts
import OpenAI from "openai";
import { config } from "../config.js";
import { IIssue } from "../models/index.js";

const openai = config.openaiApiKey ? new OpenAI({ apiKey: config.openaiApiKey }) : null;

type LintPattern = [RegExp, IIssue["severity"], string, string];

const LINT_PATTERNS: Record<string, LintPattern[]> = {
  python: [
    [/^\s*print\(/, "info", "style", "Use logging instead of print() for production code"],
    [/except\s*:/, "warning", "best-practice", "Bare except clause — catch specific exceptions"],
    [/^\s*pass\s*$/, "info", "style", "Unimplemented block — remove or implement pass"],
    [/from\s+\w+\s+import\s+\*/, "error", "linting", "Wildcard imports pollute namespace — import explicitly"],
    [/\t+/, "warning", "style", "Tabs used for indentation — use spaces (PEP 8)"],
  ],
  javascript: [
    [/console\.log\(/, "info", "style", "Remove console.log in production code"],
    [/\bvar\s+/, "warning", "style", "Use let/const instead of var"],
    [/==\s[^=]/, "warning", "best-practice", "Use === instead of =="],
    [/\beval\s*\(/, "error", "security", "eval() is dangerous — never use it"],
    [/document\.write\(/, "warning", "security", "Avoid document.write() — use DOM methods"],
  ],
  typescript: [
    [/console\.log\(/, "info", "style", "Remove console.log in production code"],
    [/: any\b/, "warning", "style", "Avoid 'any' — use proper types"],
    [/@ts-ignore/, "warning", "best-practice", "Use @ts-expect-error instead of @ts-ignore"],
    [/\beval\s*\(/, "error", "security", "eval() is dangerous — never use it"],
  ],
  go: [
    [/error\.New\(fmt\.Sprintf/, "info", "style", "Use fmt.Errorf() instead of error.New(fmt.Sprintf())"],
    [/\bpanic\(/, "warning", "best-practice", "Avoid panic outside init/main — return errors"],
    [/defer\s+os\.Remove/, "info", "linting", "Consider checking Remove error even when deferred"],
  ],
  rust: [
    [/\.unwrap\(\)/, "warning", "best-practice", "Avoid unwrap() — handle errors with match or ?"],
    [/\/\/\s*TODO/, "info", "style", "Incomplete TODO — address before release"],
    [/unsafe\s*\{/, "warning", "security", "Unsafe block — verify memory safety"],
  ],
  java: [
    [/System\.out\.print(ln)?\(/, "info", "style", "Use a logger instead of System.out"],
    [/catch\s*\(Exception\s+\w+\)/, "warning", "best-practice", "Catching generic Exception — be more specific"],
    [/public\s+static\s+void\s+main/, "info", "linting", "Verify main method placement"],
  ],
};

export const LANGUAGE_OPTIONS = [
  { id: "python", label: "Python" },
  { id: "javascript", label: "JavaScript" },
  { id: "typescript", label: "TypeScript" },
  { id: "go", label: "Go" },
  { id: "rust", label: "Rust" },
  { id: "java", label: "Java" },
];

function runLocalLint(code: string, language: string): IIssue[] {
  const issues: IIssue[] = [];
  const patterns = LINT_PATTERNS[language] || [];
  const lines = code.split("\n");

  for (let i = 0; i < lines.length; i++) {
    for (const [regex, severity, category, message] of patterns) {
      if (regex.test(lines[i])) {
        issues.push({ line: i + 1, severity, category, message });
      }
    }
  }
  return issues;
}

function buildPrompt(code: string, language: string, lintIssues: IIssue[]): string {
  const lintSection = lintIssues.length
    ? lintIssues.map((i) => `- Line ${i.line} [${i.severity}] [${i.category}]: ${i.message}`).join("\n")
    : "No basic lint issues detected.";

  return `You are a senior code reviewer. Analyze the following ${language} code and provide:

1. **Summary** — 2-3 sentence overview of the code quality
2. **Overall Score** — integer from 0 to 100
3. **Additional Issues** — any issues NOT already listed (security, performance, bugs, design, edge cases)

Existing lint issues found:
${lintSection}

Code to review:
\`\`\`${language}
${code}
\`\`\`

Respond with valid JSON ONLY (no markdown, no code fences):
{
  "summary": "...",
  "overall_score": 0-100,
  "additional_issues": [
    {"line": <int or null>, "severity": "error|warning|info", "category": "security|performance|bug|design|best-practice", "message": "...", "suggestion": "..."}
  ]
}`;
}

async function callLLM(prompt: string): Promise<any | null> {
  if (!openai) return null;
  try {
    const response = await openai.chat.completions.create({
      model: config.openaiModel,
      messages: [
        { role: "system", content: "You are a JSON-only code reviewer." },
        { role: "user", content: prompt },
      ],
      temperature: 0.2,
      max_tokens: 2000,
    });
    const text = response.choices[0]?.message?.content || "";
    const cleaned = text.replace(/```json\s*/gi, "").replace(/```\s*/g, "").trim();
    return JSON.parse(cleaned);
  } catch {
    return null;
  }
}

function computeScore(lintIssues: IIssue[], aiIssues: IIssue[]): number {
  let base = 100;
  for (const issue of [...lintIssues, ...aiIssues]) {
    if (issue.severity === "error") base -= 15;
    else if (issue.severity === "warning") base -= 8;
    else base -= 3;
  }
  return Math.max(base, 0);
}

export async function reviewCode(code: string, language: string): Promise<{
  issues: IIssue[];
  summary: string;
  score: number;
}> {
  const lintIssues = runLocalLint(code, language);
  const allIssues = [...lintIssues];
  let summary = "";
  let score = computeScore(lintIssues, []);

  const aiResult = await callLLM(buildPrompt(code, language, lintIssues));

  if (aiResult) {
    summary = aiResult.summary || "";
    const additional: any[] = aiResult.additional_issues || [];
    for (const iss of additional) {
      allIssues.push({
        line: iss.line || 0,
        severity: iss.severity || "info",
        category: iss.category || "best-practice",
        message: iss.message || "",
        suggestion: iss.suggestion || null,
      });
    }
    score = aiResult.overall_score ?? computeScore(lintIssues, additional);
  } else {
    summary = `Reviewed ${lintIssues.length} issue(s) via static analysis. Set OPENAI_API_KEY for AI-powered review.`;
  }

  return { issues: allIssues, summary, score };
}
